### PR TITLE
feat(checkbox): component can be used outside ion-item

### PR DESCRIPTION
**Pull request by amandaejohnston**
## Pull request checklist

Please check if your PR fulfills the following requirements:
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been reviewed and added / updated if needed (for bug fixes / features)
  - Some docs updates need to be made in the `ionic-docs` repo, in a separate PR. See the [contributing guide](https://github.com/ionic-team/ionic-framework/blob/main/.github/CONTRIBUTING.md#modifying-documentation) for details.
- [x] Build (`npm run build`) was run locally and any changes were pushed
- [x] Lint (`npm run lint`) has passed locally and any fixes were made for failures

## Pull request type

 

 

Please check the type of change your PR introduces:
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 

## What is the current behavior?

Issue URL: N/A

## What is the new behavior?

`ion-checkbox` can now be used outside `ion-item`, using the new default slot to specify a label.

This PR is a combination of the following:
- https://github.com/ionic-team/ionic-framework/pull/26438
- https://github.com/ionic-team/ionic-framework/pull/26460
- https://github.com/ionic-team/ionic-framework/pull/26467
- https://github.com/ionic-team/ionic-framework/pull/26471

See those individual descriptions for more details.

## Does this introduce a breaking change?

- [x] Yes
- [ ] No

The `--background` and `--background-checked` CSS properties have been renamed to `--checkbox-background` and `--checkbox-background-checked` respectively.

### CODE REVIEW

File: core/src/components/checkbox/test/a11y/index.html [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    <script type="module" src="../../../../../dist/ionic/ionic.esm.js"></script>
+  </head>
+
+  <body>
+    <main>
+      <h1>Checkbox - a11y</h1>
+
+      <ion-checkbox>Label</ion-checkbox><br />
+      <ion-checkbox aria-label="my aria label"></ion-checkbox>
+    </main>
```
Comment (liamdebeasi) [inline core/src/components/checkbox/test/a11y/index.html:20]: Can we add one with the checkbox in an item (using the `label` property)?

File: core/src/components/checkbox/checkbox.scss [hunk 2]
```diff
@@ -50,30 +60,90 @@ label {
...
+.checkbox-wrapper {
+  display: flex;
+
+  flex-grow: 1;
+
+  align-items: center;
+
+  height: 100%;
+  
+  cursor: inherit;
```
Comment (liamdebeasi) [inline core/src/components/checkbox/checkbox.scss:72]: Checkbox does not have `cursor: pointer` set on the host, so there's no indication that it is interactive. The legacy checkbox appears to place this on the invisible `label` element, but we may want to do this on the host so developers can easily override it.

File: core/src/components/checkbox/checkbox.scss [hunk 3]
```diff
@@ -50,30 +60,90 @@ label {
...
+}
+
+/**
+ * Allows us to set the background of the container
+ * While keeping the color and opacity separate.
+ */
+.checkbox-wrapper::after {
+  @include button-state();
+    
+  z-index: -1;
```
Comment (liamdebeasi) [inline core/src/components/checkbox/checkbox.scss:82]: I'm not sure I understand the comment. Can you clarify?

File: core/src/components/checkbox/test/basic/checkbox.e2e.ts [hunk 4]
```diff
@@ -5,57 +5,58 @@ test.describe('checkbox: basic', () => {
   test('should not have visual regressions', async ({ page }) => {
     await page.goto(`/src/components/checkbox/test/basic`);
 
-    await page.setIonViewport();
-
-    expect(await page.screenshot()).toMatchSnapshot(`checkbox-basic-${page.getSnapshotSettings()}.png`);
+    const content = page.locator('#checkboxes');
```
Comment (liamdebeasi) [inline core/src/components/checkbox/test/basic/checkbox.e2e.ts:8]: Can we also have a `color` test similar to https://github.com/ionic-team/ionic-framework/blob/feature-7.0/core/src/components/toggle/test/color/toggle.e2e.ts?

File: core/src/components/checkbox/test/basic/checkbox.e2e.ts [hunk 5]
```diff
@@ -5,57 +5,58 @@ test.describe('checkbox: basic', () => {
...
+    expect(await content.screenshot()).toMatchSnapshot(`checkbox-basic-${page.getSnapshotSettings()}.png`);
```
Comment (liamdebeasi) [inline core/src/components/checkbox/test/basic/checkbox.e2e.ts:9]: Can we break these out into separate tests so it's clear what each checkbox is designed to test? For example, I have a `states` test for toggle where each screenshot tests a specific state: https://github.com/ionic-team/ionic-framework/blob/feature-7.0/core/src/components/toggle/test/states/toggle.e2e.ts

File: core/src/components/checkbox/test/indeterminate/checkbox.e2e.ts [hunk 6]
```diff
@@ -2,11 +2,12 @@ import { expect } from '@playwright/test';
...
-    await page.goto(`/src/components/checkbox/test/indeterminate`);
+  test('should not have visual regressions', async ({ page, skip }) => {
+    skip.rtl();
 
-    await page.setIonViewport();
+    await page.goto(`/src/components/checkbox/test/indeterminate`);
 
-    expect(await page.screenshot()).toMatchSnapshot(`checkbox-indeterminate-${page.getSnapshotSettings()}.png`);
+    const content = page.locator('#checkboxes');
+    expect(await content.screenshot()).toMatchSnapshot(`checkbox-indeterminate-${page.getSnapshotSettings()}.png`);
```
Comment (liamdebeasi) [inline core/src/components/checkbox/test/indeterminate/checkbox.e2e.ts:11]: Do we need to check the colors here, or could that be handled by the `color` test? (see my other comment re: the `color` test)

Also, can we take a screenshot of the individual checkbox?

**Review by liamdebeasi — CHANGES_REQUESTED** [review]
Can we create a ticket and add any TODOs to the codebase so we know what to clean up when we remove the legacy checkbox?

File: core/src/components/checkbox/checkbox.scss (hunk 3 above)
Comment (amandaejohnston) [reply core/src/components/checkbox/checkbox.scss:82]: Ah whoops, this isn't actually needed anymore. It was a leftover to enable the new background CSS properties, but we've removed those. Thanks for the callout!

**Review by liamdebeasi — APPROVED** [review]
Great job!

File: core/src/components/checkbox/checkbox.scss [hunk 7]
```diff
@@ -21,26 +21,42 @@
...
 }
 
+// TODO(FW-3100): remove this
+:host(.legacy-checkbox) {
+  width: var(--size);
+  height: var(--size);
+}
+
+// TODO(FW-3100): remove :not selector
+:host(.in-item:not(.legacy-checkbox)) {
```
Comment (sean-perkins) [inline core/src/components/checkbox/checkbox.scss:47]: If we reordered the styles, wouldn't that avoid the `:not` selector based on style priority? 

e.g.:

```css
:host(.in-item) {
  width: 100%;
  height: 100%;
}

:host(.legacy-checkbox) {
  width: var(--size);
  height: var(--size);
}
```

Or am I misunderstanding what this style block is used for?

File: core/src/components/checkbox/checkbox.scss [hunk 8]
```diff
@@ -50,30 +66,82 @@ label {
...
+
+  height: 100%;
+  
+  cursor: inherit;
+}
+
+.label-text-wrapper {
+  /**
+   * This ensures that double tapping this text
+   * clicks the <label> and focuses the input
```
Comment (sean-perkins) [inline core/src/components/checkbox/checkbox.scss:84]:
```suggestion
   * clicks the <label> and focuses the checkbox
```

File: core/src/components/checkbox/checkbox.tsx [hunk 9]
```diff
@@ -143,27 +203,105 @@ export class Checkbox implements ComponentInterface {
...
-      path = indeterminate ? (
-        <path d="M2 12H22" part="mark" />
-      ) : (
-        <path d="M1.73,12.91 8.1,19.28 22.79,4.59" part="mark" />
+  // TODO(FW-3100): remove this
+  private renderLegacyCheckbox() {
+    if (!this.hasLoggedDeprecationWarning) {
+      printIonWarning(
+        `Using ion-checkbox with an ion-label has been deprecated. To migrate, remove the ion-label and pass your label directly into ion-checkbox instead.
+Example: <ion-checkbox>Email:</ion-checkbox>
```
Comment (sean-perkins) [inline core/src/components/checkbox/checkbox.tsx:282]: Should we update the label example here?

e.g.:
```
Example: <ion-checkbox>My Label:</ion-checkbox>
```

**Review by sean-perkins — APPROVED** [review]
Great work! None of my comments are blocking.

File: core/src/components/checkbox/checkbox.scss (hunk 7 above)
Comment (amandaejohnston) [reply core/src/components/checkbox/checkbox.scss:47]: Aha, good thinking 👀

**Pull request merged**